File: cli/drand_draw/lottery.py
import math


def compute_round(deadline, genesis, period):
    elapsed = deadline - genesis
    if elapsed < 0:
        return -1
    return math.floor(elapsed / period) + 1


def compute_winners(randomness, n, prize_tiers):
    total = sum(prize_tiers)
    if total > n:
        raise ValueError(f"Total prizes ({total}) exceeds N ({n})")
    winners = []
    used = set()
    for count in prize_tiers:
        for _ in range(count):
            shift = len(winners)
            seed_hex = randomness + format(shift, 'x')
            big_val = int(seed_hex, 16)
            idx = big_val % n
            attempts = 0
            while idx in used:
                idx = (idx + 1) % n
                attempts += 1
                if attempts >= n:
                    raise RuntimeError("Collision resolution failed")
            used.add(idx)
            winners.append(idx)
    return winners

File: src/lottery.js
import { CHAINS } from './chains.js'

export function computeRound(chainId, deadlineUnix) {
  const chain = CHAINS[chainId]
  const elapsed = deadlineUnix - chain.genesisTime
  if (elapsed < 0) return -1
  return Math.floor(elapsed / chain.period) + 1
}

async function deriveSeed(randomness, shift) {
  const input = randomness + ':' + shift
  const enc = new TextEncoder().encode(input)
  const hash = await crypto.subtle.digest('SHA-256', enc)
  const hex = Array.from(new Uint8Array(hash)).map(b => b.toString(16).padStart(2, '0')).join('')
  return hex
}

export async function computeWinners(randomness, n, prizeTiers) {
  const totalPrizes = prizeTiers.reduce((a, b) => a + b, 0)
  if (totalPrizes > n) {
    throw new Error(`Total prizes (${totalPrizes}) exceeds N (${n})`)
  }
  const winners = []
  const used = new Set()
  for (let i = 0; i < prizeTiers.length; i++) {
    for (let j = 0; j < prizeTiers[i]; j++) {
      const shift = winners.length
      const seedHex = await deriveSeed(randomness, shift)
      const bigVal = BigInt('0x' + seedHex)
      let idx = Number(bigVal % BigInt(n))
      let attempts = 0
      while (used.has(idx)) {
        idx = (idx + 1) % n
        attempts++
        if (attempts >= n) throw new Error('Collision resolution failed')
      }
      used.add(idx)
      winners.push(idx)
    }
  }
  return winners
}

export function formatWinnerList(winners, prizeTiers) {
  const result = []
  let idx = 0
  for (let i = 0; i < prizeTiers.length; i++) {
    const tierWinners = winners.slice(idx, idx + prizeTiers[i])
    result.push(tierWinners)
    idx += prizeTiers[i]
  }
  return result
}
